std/range/package.d

```diff
@@ -8247,40 +8249,36 @@ unittest
     auto cWrapper = refRange(&c);
     static assert(is(typeof(cWrapper) == C));
     assert(cWrapper is c);
+}
 
-    struct S
+unittest // issue 14373
+{
+    static struct R
     {
-        @property int front() @safe const pure nothrow { return 0; }
-        @property bool empty() @safe const pure nothrow { return false; }
-        void popFront() @safe pure nothrow { }
-
-        int i = 27;
+        @property int front() {return 0;}
+        void popFront() {empty = true;}
+        bool empty = false;
     }
-    static assert(isInputRange!S);
-    static assert(!isForwardRange!S);
-
-    auto s = S(42);
-    auto sWrapper = refRange(&s);
-    static assert(is(typeof(sWrapper) == S));
-    assert(sWrapper == s);
+    R r;
+    refRange(&r).popFront();
+    assert(r.empty);
 }
 
 /++
     Helper function for constructing a $(LREF RefRange).
 
-    If the given range is not a forward range or it is a class type (and thus is
-    already a reference type), then the original range is returned rather than
-    a $(LREF RefRange).
+    If the given range is a class type (and thus is already a reference type),
+    then the original range is returned rather than a $(LREF RefRange).
   +/
 auto refRange(R)(R* range)
-    if(isForwardRange!R && !is(R == class))
+    if(!is(R == class))
```

Comment:
if(isInputRange!R && !is(R == class)) 
or 
static assert(isInputRange!R, "refRange only works with ranges");

Wrong types should be detected as early as possible.